File: macos/Services/MCP/Tools/Tier1/ListSchemasTool.swift
// ListSchemasTool.swift
// Gridex
//
// MCP Tool: List schemas/databases available in a connection.

import Foundation

struct ListSchemasTool: MCPTool {
    let name = "list_schemas"
    let description = "List schemas/databases available in a connection."
    let tier = MCPPermissionTier.schema

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ]
        ],
        "required": ["connection_id"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)

        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        let (adapter, config) = try await context.getAdapter(for: connectionId)

        // For databases that support schemas (PostgreSQL, MSSQL)
        if config.databaseType.supportsSchemas {
            let schemas = try await adapter.listSchemas(database: nil)
            if schemas.isEmpty {
                return MCPToolResult(text: "No schemas found.")
            }
            let json = try JSONEncoder().encode(schemas)
            return MCPToolResult(text: "Schemas: " + (String(data: json, encoding: .utf8) ?? "[]"))
        }

        // For MySQL and others, list databases
        let databases = try await adapter.listDatabases()
        if databases.isEmpty {
            return MCPToolResult(text: "No databases found.")
        }
        let json = try JSONEncoder().encode(databases)
        return MCPToolResult(text: "Databases: " + (String(data: json, encoding: .utf8) ?? "[]"))
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier1/ListSchemasTool.cpp
//
// ListSchemasTool.cpp
//

#include "ListSchemasTool.h"
#include "../MCPToolHelpers.h"

namespace DBModels
{
    static bool supportsSchemas(DatabaseType t)
    {
        return t == DatabaseType::PostgreSQL || t == DatabaseType::MSSQLServer;
    }

    MCPToolResult ListSchemasTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);

        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        auto [adapter, config] = ctx.getAdapter(connId);

        nlohmann::json arr = nlohmann::json::array();
        std::wstring kind = L"Schemas";
        if (supportsSchemas(config.databaseType))
        {
            for (const auto& s : adapter->listSchemas())
                arr.push_back(MCPToolHelpers::toUtf8(s));
        }
        else
        {
            kind = L"Databases";
            for (const auto& d : adapter->listDatabases())
                arr.push_back(MCPToolHelpers::toUtf8(d));
        }

        if (arr.empty())
            return MCPToolResult::text("No " + MCPToolHelpers::toUtf8(kind) + " found.");

        return MCPToolResult::text(
            MCPToolHelpers::toUtf8(kind) + ": " + arr.dump());
    }
}
